cpp/inline/games/nim/Game.inl

```diff
@@ -0,0 +1,98 @@
+#include <games/nim/Game.hpp>
+
+namespace nim {
+
+inline void Game::Rules::init_state(State& state) {
+  state.stones_left = kStartingStones;
+  state.current_player = 0;
+}
+
+inline size_t Game::State::hash() const {
+  auto tuple = std::make_tuple(stones_left, current_player, chance_active);
+  std::hash<decltype(tuple)> hasher;
+  return hasher(tuple);
+}
+
+inline Game::Types::ActionMask Game::Rules::get_legal_moves(const StateHistory& history) {
+  const State& state = history.current();
+  Types::ActionMask mask;
+  bool is_chance = is_chance_mode(get_action_mode(history.current()));
+  if (is_chance) {
+    for (int i = 0; i < std::min(nim::kChanceDistributionSize, state.stones_left + 1); ++i) {
+      mask[i] = true;
+    }
+  } else {
+    for (int i = 0; i < std::min(nim::kMaxStonesToTake, state.stones_left); ++i) {
+      mask[i] = true;
+    }
+  }
+
+  return mask;
+}
+
+// current_player only switches AFTER a chance action
+inline void Game::Rules::apply(StateHistory& history, core::action_t action) {
+  bool is_chance = is_chance_mode(get_action_mode(history.current()));
+  State& state = history.extend();
+
+  if (is_chance) {
+    int outcome_stones = state.stones_left - action;
+    state.stones_left = outcome_stones;
+    state.current_player = 1 - state.current_player;
+    state.chance_active = false;
+  } else {
+    if (action < 0 || action >= nim::kMaxStonesToTake) {
+      throw std::invalid_argument("Invalid action: " + std::to_string(action));
+    }
+    state.stones_left = state.stones_left- (action + 1);
+    state.chance_active = true;
+  }
+}
+
+// if the game ends after a chance action, the player who made the last move wins
+inline bool Game::Rules::is_terminal(const State& state, core::seat_index_t last_player,
+                                     core::action_t last_action, GameResults::Tensor& outcome) {
+  if (state.stones_left == 0) {
+    outcome.setZero();
+    outcome(last_player) = 1;
+    return true;
+  }
+  return false;
+}
+
+inline Game::Types::ChanceDistribution Game::Rules::get_chance_distribution(const State& state) {
+  if (!is_chance_mode(get_action_mode(state))) {
+    throw std::invalid_argument("Not in chance mode");
+  }
+
+  Game::StateHistory single_state_history;
+  single_state_history.update(state);
+  int num_legal_moves = get_legal_moves(single_state_history).count();
+
+  Types::ChanceDistribution dist;
+  dist.setZero();
+  for (int i = 0; i < num_legal_moves; ++i) {
+    dist(i) = nim::kChanceEventProbs[i];
+  }
+  for (int i = num_legal_moves; i < nim::kChanceDistributionSize; ++i) {
+    dist(num_legal_moves - 1) += nim::kChanceEventProbs[i];
+  }
```

Comment:
This second for-loop deserves an explanatory comment.

In my opinion, your intention is clearer if you replace the 2 for-loops with:

```
float cumulative_prob = 0;
for (int i = 0; i < num_stones; ++i) {
  dist(i) = nim::kChanceEventProbs[i];
  cumulative_prob += dist(i);
}
dist(num_stones) = 1.0 - cumulative_prob;
```

But still deserves a comment.